scala-api/src/test/scala/com/cloudera/livy/scalaapi/ScalaClientTest.scala

```diff
@@ -0,0 +1,243 @@
+    val sFuture = client.submit(ScalaClientTest.simpleSparkJob)
+    val result = Await.result(sFuture, Timeout second)
+    assert(result === 5)
+  }
+
+  test("test Job Failure") {
+    configureClient(true)
+    val sFuture = client.submit(ScalaClientTest.throwExceptionJob)
+    val lock = new CountDownLatch(1)
+    var testFailure : Option[String] = None
+    sFuture onComplete {
+      case Success(t) => {
+        testFailure = Some("Test should have thrown CustomFailureException")
+        lock.countDown()
+      }
+      case Failure(e) => {
+        if (!e.getMessage.contains("CustomTestFailureException")) {
+          testFailure = Some("Test did not throw expected exception - CustomFailureException")
+        }
+        lock.countDown()
+      }
+    }
+    if (lock.getCount == 1) lock.await()
+    testFailure.foreach(fail(_))
+  }
+
+  test("test Sync Rpc") {
+    configureClient(true)
+    val future = client.run(ScalaClientTest.helloJob)
+    val result = Await.result(future, Timeout second)
+    assert(result === "hello")
+  }
+
+  test("test Remote client") {
+    configureClient(false)
+    val sFuture = client.submit(ScalaClientTest.simpleSparkJob)
+    val result = Await.result(sFuture, Timeout second)
+    assert(result === 5)
+  }
+
+  test("test add file") {
+    configureClient(true)
+    val file = File.createTempFile("test", ".file")
+    val fileStream = new FileOutputStream(file)
+    fileStream.write("test file".getBytes("UTF-8"))
+    fileStream.close
+    val addFileFuture = client.addFile(new URI("file:" + file.getAbsolutePath()))
+    Await.ready(addFileFuture, Timeout second)
+    val sFuture = client.submit { context =>
+      ScalaClientTest.fileOperation(false, file.getName, context)
+    }
+    val output = Await.result(sFuture, Timeout second)
+    assert(output === "test file")
+  }
+
+  test("test add jar") {
+    configureClient(true)
+    val jar = File.createTempFile("test", ".resource")
+    val jarFile = new JarOutputStream(new FileOutputStream(jar))
+    jarFile.putNextEntry(new ZipEntry("test.resource"))
+    jarFile.write("test resource".getBytes("UTF-8"))
+    jarFile.closeEntry()
+    jarFile.close()
+    val addJarFuture = client.addJar(new URI("file:" + jar.getAbsolutePath()))
+    Await.ready(addJarFuture, Timeout second)
+    val sFuture = client.submit { context =>
+      ScalaClientTest.fileOperation(true, "test.resource", context)
+    }
+    val output = Await.result(sFuture, Timeout second)
+    assert(output === "test resource")
+  }
+
+  test("Successive onComplete callbacks") {
+    var testFailure: Option[String] = None
+    configureClient(true)
+    val future = client.run(ScalaClientTest.helloJob)
+    var count: Integer = 0
+    breakable {
+      for (i <- 0 to 2) {
+        val lock = new CountDownLatch(1)
+        future onComplete {
+          case Success(t) => {
+            if (!t.equals("hello")) testFailure = Some("Expected message not returned")
```

Comment:
nit: expand blocks like this.

```
if (condition) {
  doSomething()
}
```

The extra verbosity helps with readability.